timemachine/fe/single_topology.py

```diff
@@ -373,38 +405,37 @@ def setup_end_state(ff, mol_a, mol_b, core, a_to_c, b_to_c):
-        [canonicalize_chiral_atom_idxs(x) for x in mol_a_chiral_atom_idxs]
-        + [canonicalize_chiral_atom_idxs(x) for x in all_dummy_chiral_atom_idxs]
-    )
-
-    # all chiral atom interactions present in mol_b, *excluding* dummy group interactions from mol_a
-    canon_mol_b_chiral_atom_idxs = set(
-        [canonicalize_chiral_atom_idxs(x) for x in recursive_map(mol_b_chiral_atom.potential.idxs, b_to_c)]
-    )
-
-    chiral_atom_idxs_turned_off_at_end_state = canon_mol_b_chiral_atom_idxs.difference(canon_mol_a_chiral_atom_idxs)
-    all_angles_implied_by_turned_off_chiral_atom_idxs = set()
-    for dummy_center, dummy_i, dummy_j, dummy_k in chiral_atom_idxs_turned_off_at_end_state:
-        all_angles_implied_by_turned_off_chiral_atom_idxs.add(canonicalize_bond([dummy_i, dummy_center, dummy_j]))
-        all_angles_implied_by_turned_off_chiral_atom_idxs.add(canonicalize_bond([dummy_i, dummy_center, dummy_k]))
-        all_angles_implied_by_turned_off_chiral_atom_idxs.add(canonicalize_bond([dummy_j, dummy_center, dummy_k]))
-
-    all_dummy_angle_params_modified = []
-    for angle_idxs, angle_params in zip(all_dummy_angle_idxs, all_dummy_angle_params):
-        if angle_idxs in all_angles_implied_by_turned_off_chiral_atom_idxs:
-            all_dummy_angle_params_modified.append([MINIMUM_CHIRAL_ANGLE_FORCE_CONSTANT, angle_params[1]])
-        else:
-            all_dummy_angle_params_modified.append(angle_params)
+    # process chiral volumes, turning off ones at the end-state that have a missing bond.
+    canon_mol_a_bond_idxs_set = set([canonicalize_bond(x) for x in mol_a_bond_idxs])
+    # assert presence of bonds
+    for c, i, j, k in mol_a_chiral_atom_idxs:
+        ci = canonicalize_bond((c, i))
+        cj = canonicalize_bond((c, j))
+        ck = canonicalize_bond((c, k))
+        assert ci in canon_mol_a_bond_idxs_set
+        assert cj in canon_mol_a_bond_idxs_set
+        assert ck in canon_mol_a_bond_idxs_set
+
+    canon_mol_c_bond_idxs_set = set([canonicalize_bond(x) for x in mol_c_bond_idxs])
+
+    # a proper dummy chiral atom restraint c,i,j,k requires that all bonds
+    # ci, cj, ck be present at the end-state. Otherwise, it is disabled.
+    all_proper_dummy_chiral_atom_idxs = []
+    all_proper_dummy_chiral_atom_params = []
+    for (c, i, j, k), p in zip(all_dummy_chiral_atom_idxs, all_dummy_chiral_atom_params):
+        ci = canonicalize_bond((c, i))
+        cj = canonicalize_bond((c, j))
+        ck = canonicalize_bond((c, k))
+        if ci in canon_mol_c_bond_idxs_set and cj in canon_mol_c_bond_idxs_set and ck in canon_mol_c_bond_idxs_set:
+            all_proper_dummy_chiral_atom_idxs.append((c, i, j, k))
+            all_proper_dummy_chiral_atom_params.append(p)
+        # leave the below for debugging
+        #     print("Chiral Volume", c,i,j,k, "is okay, turning on.")
+        # else:
+        #     print("Chiral Volume", c,i,j,k, "has a disabled bond, turning off.")
```

Comment:
This seems like it might be useful to have as a warning

Reply:
willfix